sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -118,6 +118,8 @@ abstract class Optimizer(sessionCatalog: SessionCatalog, conf: CatalystConf)
       SimplifyCreateMapOps) ::
     Batch("Check Cartesian Products", Once,
       CheckCartesianProducts(conf)) ::
+    Batch("Join Reorder", Once,
+      CostBasedJoinReorder(conf)) ::
```

Comment:
CostBasedJoinReorder can create two consecutive projections. The top projection will have some expressions, and the bottom projection will only have attributes (it is created by the join reorder rule). We can get rid of these stacked projections by calling `CollapseProject` rule after the reordering.